### PR TITLE
FormTokenField: Add Help Text

### LINKED ISSUE #15355: Useful content is unnecessarily hidden
Useful content is unnecessarily hidden

-   **Severity:** Low
-   **Affected Populations:**
    -   Motor Impaired
    -   Cognitively Impaired
-   **Platform(s):**
    -   All / Universal
-   **Components affected:**
    -   Block Panel

#### Issue description

After the tags input, where users can add tags to their document, hidden
text states that these can be comma-separated. This text is only
available to screen reader users, even though these instructions are
useful to many sighted users, such as those with a cognitive disability
and keyboard-only users.

##### Issue Code

```
    <div class="components-form-token-field" tabindex="-1">


        <label for="..." class="...">Add New Tag</label>


        <div class="..." tabindex="-1">


            <input id="..." type="text" size="1" class="..." role="combobox" aria-expanded="false" aria-autocomplete="list" aria-describedby="..." value="">


        </div>


        <div id="..." class="screen-reader-text">Separate with commas</div>


    </div>
```

#### Remediation Guidance

Allow the instructions to be available to all users, not just screen
reader users.

##### Recommended Code

```
    <div class="components-form-token-field" tabindex="-1">


        <label for="..." class="...">Add New Tag</label>


        <div class="..." tabindex="-1">


            <input id="..." type="text" size="1" class="..." role="combobox" aria-expanded="false" aria-autocomplete="list" aria-describedby="..." value="">


        </div>


        <div id="...">Separate with commas</div>


    </div>
```

**Note:** This issue may be a duplicate with other existing accessibility-related bugs in this project. This issue comes from the Gutenberg accessibility audit, performed by [Tenon](https://www.tenon.io) and funded by [WP Campus](https://wpcampus.org/). This issue is GUT-40 in Tenon's report

### PR DESCRIPTION
**Pull request by marekhrabe**
## Description
Fixes #15355

I've made the message visible and added the mention of the enter key. It's styled to look the same way as other help texts in the sidebar.

The component allows developers to opt-in for using spacebar as a separator too. I've reflected that and made two versions of the help text based on the configuration.

## How has this been tested?
- New Post
- In the sidebar, open "Tags" Panel and explore the new message

The sidebar term selector (Tags, Categories) is the only place where this component is used in core.

## Screenshots

| Standard Usage | With Spacebar Separator Enabled |
| - | - |
| Screenshot 2019-05-06 at 17 49 59 | Screenshot 2019-05-06 at 17 51 16

### CODE REVIEW

File: packages/components/src/form-token-field/style.scss
Comment (afercia) [inline packages/components/src/form-token-field/style.scss]: This makes the label extend to all the available width, basically an area without text becomes clickable which should be avoided and the reason why `inline-block` was used. Unless I'm missing something, please revert.

File: packages/components/src/form-token-field/style.scss
Comment (afercia) [inline packages/components/src/form-token-field/style.scss]: This applies to a `<div>` element, which is a block element by default. Unless I'm missing something, please remove this property. Also, I'd make this a `<p>` element, see comment above.

File: packages/components/src/form-token-field/index.js
Comment (afercia) [inline packages/components/src/form-token-field/index.js]: Since we're touching this part, I'd change this `<div>` to a `<p>`: slightly better from a semantic perspective.

File: packages/components/src/form-token-field/style.scss
Comment (afercia) [inline packages/components/src/form-token-field/style.scss]: I'd like to have some design feedback about the italic style. Personally, I'd avoid it because some users find it difficult to read.

File: packages/components/src/form-token-field/index.js
Comment (afercia) [inline packages/components/src/form-token-field/index.js]: Nice catch on differentiating the messages! I guess "Enter" should have a capital E. I believe there's the need for an Oxford comma somewhere :)

**Review by afercia — CHANGES_REQUESTED** [review]
Thanks @marekhrabe, I left some comments.

Re: the Italic text, I'd like some design feedback. I'd like to avoid it as much as possible for better readability. Also, seems to me Gutenberg isn't using Italic for this kind of hints underneath input fields, for example:

Screenshot 2019-05-06 at 18 24 23

/Cc @karmatosed

File: packages/components/src/form-token-field/index.js
Comment (jorgefilipecosta) [inline packages/components/src/form-token-field/index.js]: It seems like the id `components-form-token-suggestions-howto-${ instanceId }` is not being referenced anywhere, can we remove the id?

**Review by jorgefilipecosta — COMMENTED** [review]
It seems this component has a label and a help text feature offered by the BaseControl component: https://github.com/WordPress/gutenberg//blob/1e40b281d1db725e512e383b5e00f3b59e333d5a/packages/components/src/base-control/README.md

Would it be possible to use the base control here, avoid the need for label and help text custom styles and promoting code reusability?

**Comment by marekhrabe** [comment]
I've actually tried using BaseControl but it cannot be easily used as this component attaches a lot of additional attributes on the wrapping element and that's not supported.

File: packages/components/src/form-token-field/style.scss
Comment (marekhrabe) [reply packages/components/src/form-token-field/style.scss]: I have made it italic based on styling of the help text of `BaseControl` which is used as a base for most of other controls. If there are italics missing for help texts similar to this, they are inconsistent with the rest. Just a few examples:

Screenshot 2019-05-06 at 21 18 19
Screenshot 2019-05-06 at 21 18 29
Screenshot 2019-05-06 at 21 19 10

I agree on the issue with italics but I think we should address that separately and on a global level for all those existing usages of BaseControl. What do you think about leaving this PR as italics and opening a discussion to replace them globally for all help texts? I'll be happy to make that PR once we settle on a solution.

File: packages/components/src/form-token-field/index.js
Comment (marekhrabe) [reply packages/components/src/form-token-field/index.js]: Referenced here https://github.com/WordPress/gutenberg/blob/master/packages/components/src/form-token-field/token-input.js#L50

File: packages/components/src/form-token-field/style.scss
Comment (kjellr) [reply packages/components/src/form-token-field/style.scss]: Yeah, I think italics vs. no italics can be discussed separately — this PR is just mirroring the existing styles used in `.components-base-control__help`. 

(From some quick clicking around, it seems that the [Permalink help text](https://github.com/WordPress/gutenberg/pull/15469#pullrequestreview-234087938) is the only text of this kind that isn't italicized, so we should make a decision one way or another).

Comment (afercia) [reply packages/components/src/form-token-field/style.scss]: Italic is terrible for some kind of dyslexia when used on whole sentences. I vote for just not add new Italic. Also, planning to create related GitHub issue and Trac ticket. Thank you.

Comment (kjellr) [reply packages/components/src/form-token-field/style.scss]: Cool — I'm fine getting rid of the italic globally if we can replace it with something else. I just want to make sure that help text appears differently from the form field label so users can tell them apart. But there are other ways to do that besides making the text italic. I'll keep an eye out for the ticket and try out some ideas there. 🙂

Comment (afercia) [reply packages/components/src/form-token-field/style.scss]: > there are other ways to do that besides making the text italic
Definitely agree. One day before I die I'd like to see a typographic system in WordPress 🙂

Comment (kjellr) [reply packages/components/src/form-token-field/style.scss]: I've opened https://github.com/WordPress/gutenberg/issues/15683 to track the inconsistent help text, and to consider the use of italic text there. We'll get that sorted out globally in Gutenberg. 🙂

In the meantime, that shouldn't block this PR since it's not adding any new styles — just replicating the existing `BaseControl` help text style.

**Review by kjellr — CHANGES_REQUESTED** [review]
@marekhrabe This is looking great. Just a tiny design change before it's ready to merge on my end. Currently, the first version of the label says: 

> Separate using commas or the Enter key

We should add a period at the end there to align with other help text. Unfortunately, that awkwardly pushes the word "key." onto its own line on desktop screens. 😄 

To avoid that, let's swap out "using" with the word "with":

Screen Shot 2019-05-16 at 10 56 07 AM

We can also make that change with the second label version that includes the mention of spaces: 

Screen Shot 2019-05-16 at 1 55 31 PM

**Comment by marekhrabe** [comment]
What a fix! 😀Message updated, code rebased.

**Review by kjellr — APPROVED** [review]
Looks great from a design perspective. 👍

**Comment by kjellr** [comment]
> Re: the Italic text, I'd like some design feedback. I'd like to avoid it as much as possible for better readability.

@afercia: Now that we have https://github.com/WordPress/gutenberg/issues/15683 in progress to track this globally, is this PR all ready to go on your end?

**Comment by afercia** [comment]
👍 from me.

**Comment by marekhrabe** [comment]
Thanks, @afercia. Can you please update the review too? Github still blocks me from merging as you requested changes previously.

> Merging is blocked
> Merging can be performed automatically once the requested changes are addressed.

**Review by afercia — APPROVED** [review]

**Pull request merged**
